packages/transformer/src/IModelTransformer.ts

```diff
@@ -2939,118 +2948,103 @@ export class IModelTransformer extends IModelExportHandler {
-        // try to attach the second db and query both together anyway
-        (sourceElemFedGuid &&
-          this._queryElemIdByFedGuid(this.targetDb, sourceElemFedGuid)) ||
-        // FIXME<MIKE>: describe why it's safe to assume nothing has been deleted in provenanceDb
-        this._queryProvenanceForElement(instId);
 
-      // since we are processing one changeset at a time, we can see local source deletes
-      // of entities that were never synced and can be safely ignored
-      const deletionNotInTarget = !targetId;
-      if (deletionNotInTarget) return;
-      this.context.remapElement(instId, targetId);
-      // If an entity insert and an entity delete both point to the same entity in target iModel, that means that entity was recreated.
-      // In such case an entity update will be triggered and we no longer need to delete the entity.
-      if (alreadyImportedElementInserts.has(targetId)) {
-        this.exporter.sourceDbChanges?.element.deleteIds.delete(instId);
-      }
-      if (alreadyImportedModelInserts.has(targetId)) {
-        this.exporter.sourceDbChanges?.model.deleteIds.delete(instId);
-      }
-    } else {
-      // is deleted relationship
-      const classFullName = change.$meta?.classFullName;
-      const sourceIdOfRelationshipInSource = change.SourceECInstanceId;
-      const targetIdOfRelationshipInSource = change.TargetECInstanceId;
-      const [sourceIdInTarget, targetIdInTarget] = [
-        sourceIdOfRelationshipInSource,
-        targetIdOfRelationshipInSource,
-      ].map((id) => {
+    const changedInstanceId = change.ECInstanceId;
+    const sourceIdOfRelationshipInSource = change.SourceECInstanceId;
+    const targetIdOfRelationshipInSource = change.TargetECInstanceId;
+    const classFullName = change.$meta?.classFullName;
+    const idsInSourceDb = isRelationship
+      ? [sourceIdOfRelationshipInSource, targetIdOfRelationshipInSource]
+      : [changedInstanceId];
+    const idsInTargetDb = await Promise.all(
+      idsInSourceDb.map(async (id) => {
+        let identifierValue: string | undefined;
         let element;
-        try {
-          element = this.sourceDb.elements.getElement(id);
-        } catch (err) {
-          return undefined;
+        if (isRelationship) {
+          try {
+            element = this.sourceDb.elements.getElement(id);
+          } catch (err) {
+            return undefined;
+          }
         }
-        const fedGuid = element.federationGuid;
-        let identifierValue: string | undefined;
+        const fedGuid = isRelationship
+          ? element?.federationGuid
+          : change.FederationGuid;
         if (queryCanAccessProvenance) {
-          const aspects: ExternalSourceAspect[] =
-            this.sourceDb.elements.getAspects(
+          for await (const row of this.sourceDb.createQueryReader(
+            `SELECT esa.Identifier FROM bis.ExternalSourceAspect esa WHERE Scope.Id=:scopeId AND Kind=:kind AND Element.Id=:relatedElementId`,
```

Comment:
add a `// TODO: clarify what happens if there are multiple (e.g. elements were merged)`
also I would go with `LIMIT 1` for now

Reply:
changed it